File: src/axonize/storage.py
from __future__ import annotations

import hashlib
import json
import os
import re
import tempfile
from pathlib import Path

from axonize.models import BlockType, Edge, EdgeOrigin, FileIntegrityState, Node, RelationType


HEADING_ONLY_RE = re.compile(r"^\s{0,3}(#{1,6})\s+(.+?)\s*$", flags=re.MULTILINE)


def sha256_file(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(65536), b""):
            digest.update(chunk)
    return digest.hexdigest()


def heading_structure_hash(content: str) -> str:
    normalized = content.replace("\r\n", "\n").replace("\r", "\n")
    headings: list[str] = []
    for match in HEADING_ONLY_RE.finditer(normalized):
        level = len(match.group(1))
        title = re.sub(r"\s+", " ", match.group(2)).strip().lower()
        headings.append(f"{level}:{title}")
    digest = hashlib.sha256("\n".join(headings).encode("utf-8")).hexdigest()
    return digest


def discover_markdown_files(repo_root: Path) -> list[str]:
    docs_root = repo_root / "docs"
    if not docs_root.exists():
        return []

    files = [path for path in docs_root.rglob("*.md") if path.is_file()]
    relative = [path.relative_to(repo_root).as_posix() for path in files]
    relative.sort()
    return relative


def _read_json(path: Path) -> dict | list | None:
    if not path.exists():
        return None
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        return None


def _write_json(path: Path, payload: dict | list) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with tempfile.NamedTemporaryFile("w", encoding="utf-8", dir=path.parent, delete=False) as handle:
        json.dump(payload, handle, ensure_ascii=True, indent=2, sort_keys=True)
        handle.write("\n")
        tmp_name = handle.name
    os.replace(tmp_name, path)


def ensure_sidecar(repo_root: Path) -> Path:
    sidecar = repo_root / ".semantic"
    sidecar.mkdir(parents=True, exist_ok=True)
    (sidecar / "cache").mkdir(parents=True, exist_ok=True)
    return sidecar


def load_state(sidecar: Path) -> dict | None:
    payload = _read_json(sidecar / "state.json")
    if isinstance(payload, dict):
        return payload
    return None


def save_state(sidecar: Path, payload: dict) -> None:
    _write_json(sidecar / "state.json", payload)


def load_nodes(sidecar: Path) -> list[Node]:
    payload = _read_json(sidecar / "nodes.json")
    if not isinstance(payload, list):
        return []

    nodes: list[Node] = []
    for item in payload:
        if not isinstance(item, dict):
            continue
        try:
            nodes.append(
                Node(
                    id=str(item["id"]),
                    type=BlockType(str(item["type"])),
                    file=str(item["file"]),
                    title=str(item.get("title", "")),
                    content_hash=str(item.get("content_hash", "")),
                    structure_hash=str(item.get("structure_hash", "")),
                )
            )
        except (KeyError, ValueError):
            continue
    return nodes


def save_nodes(sidecar: Path, nodes: list[Node]) -> None:
    payload = [node.to_dict() for node in sorted(nodes, key=lambda item: (item.file, item.id))]
    _write_json(sidecar / "nodes.json", payload)


def load_edges(sidecar: Path) -> list[Edge]:
    payload = _read_json(sidecar / "edges.json")
    if not isinstance(payload, list):
        return []

    edges: list[Edge] = []
    for item in payload:
        if not isinstance(item, dict):
            continue
        try:
            relation = str(item["type"])
            if not RelationType.has_value(relation):
                continue
            origin = str(item.get("origin", EdgeOrigin.DETERMINISTIC.value))
            if origin not in {EdgeOrigin.DETERMINISTIC.value, EdgeOrigin.INFERRED.value}:
                origin = EdgeOrigin.DETERMINISTIC.value
            edges.append(
                Edge(
                    source=str(item["source"]),
                    target=str(item["target"]),
                    type=RelationType(relation),
                    weight=float(item.get("weight", 1.0)),
                    confidence=float(item.get("confidence", 1.0)),
                    origin=EdgeOrigin(origin),
                )
            )
        except KeyError:
            continue
    return edges


def save_edges(sidecar: Path, edges: list[Edge]) -> None:
    payload = [
        edge.to_dict()
        for edge in sorted(edges, key=lambda item: (item.source, item.type.value, item.target, item.origin.value))
    ]
    _write_json(sidecar / "edges.json", payload)


def build_files_state(repo_root: Path, files: list[str]) -> dict[str, FileIntegrityState]:
    state: dict[str, FileIntegrityState] = {}
    for file_path in files:
        full_path = repo_root / file_path
        content = full_path.read_text(encoding="utf-8")
        state[file_path] = FileIntegrityState(
            file_hash=sha256_file(full_path),
            structure_hash=heading_structure_hash(content),
        )
    return state

File: desktop/main/semantic/storage.ts
import { createHash } from "node:crypto";
import { promises as fs } from "node:fs";
import path from "node:path";
import { BlockType, Edge, EdgeOrigin, FileIntegrityState, Node, RelationType, isBlockType, isRelationType } from "./models";

const HEADING_ONLY_RE = /^\s{0,3}(#{1,6})\s+(.+?)\s*$/gm;

async function readJson<T>(filePath: string): Promise<T | null> {
  try {
    const raw = await fs.readFile(filePath, "utf8");
    return JSON.parse(raw) as T;
  } catch {
    return null;
  }
}

async function writeJson(filePath: string, payload: unknown): Promise<void> {
  await fs.mkdir(path.dirname(filePath), { recursive: true });
  const tempPath = `${filePath}.tmp`;
  await fs.writeFile(tempPath, `${JSON.stringify(payload, null, 2)}\n`, "utf8");
  await fs.rename(tempPath, filePath);
}

export async function ensureSidecar(repoRoot: string): Promise<string> {
  const sidecar = path.join(repoRoot, ".semantic");
  await fs.mkdir(path.join(sidecar, "cache"), { recursive: true });
  return sidecar;
}

export async function discoverMarkdownFiles(repoRoot: string): Promise<string[]> {
  const docsRoot = path.join(repoRoot, "docs");
  async function walk(dir: string): Promise<string[]> {
    let entries: Array<import("node:fs").Dirent>;
    try {
      entries = await fs.readdir(dir, { withFileTypes: true });
    } catch {
      return [];
    }

    const files: string[] = [];
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        files.push(...(await walk(fullPath)));
        continue;
      }
      if (entry.isFile() && entry.name.endsWith(".md")) {
        files.push(path.relative(repoRoot, fullPath).split(path.sep).join("/"));
      }
    }
    return files;
  }

  const files = await walk(docsRoot);
  files.sort();
  return files;
}

export async function sha256File(filePath: string): Promise<string> {
  const digest = createHash("sha256");
  const content = await fs.readFile(filePath);
  digest.update(content);
  return digest.digest("hex");
}

export function headingStructureHash(content: string): string {
  const normalized = content.replace(/\r\n/g, "\n").replace(/\r/g, "\n");
  const headings: string[] = [];
  HEADING_ONLY_RE.lastIndex = 0;
  for (const match of normalized.matchAll(HEADING_ONLY_RE)) {
    const level = (match[1] ?? "").length;
    const title = (match[2] ?? "").replace(/\s+/g, " ").trim().toLowerCase();
    headings.push(`${level}:${title}`);
  }
  return createHash("sha256").update(headings.join("\n"), "utf8").digest("hex");
}

export async function buildFilesState(repoRoot: string, files: readonly string[]): Promise<Record<string, FileIntegrityState>> {
  const state: Record<string, FileIntegrityState> = {};
  for (const filePath of files) {
    const fullPath = path.join(repoRoot, filePath);
    const content = await fs.readFile(fullPath, "utf8");
    state[filePath] = {
      file_hash: await sha256File(fullPath),
      structure_hash: headingStructureHash(content)
    };
  }
  return state;
}

export async function loadState(sidecar: string): Promise<Record<string, unknown> | null> {
  const payload = await readJson<Record<string, unknown>>(path.join(sidecar, "state.json"));
  if (payload && typeof payload === "object") {
    return payload;
  }
  return null;
}

export async function saveState(sidecar: string, payload: Record<string, unknown>): Promise<void> {
  await writeJson(path.join(sidecar, "state.json"), payload);
}

export async function loadNodes(sidecar: string): Promise<Node[]> {
  const payload = await readJson<Array<Record<string, unknown>>>(path.join(sidecar, "nodes.json"));
  if (!Array.isArray(payload)) {
    return [];
  }

  const nodes: Node[] = [];
  for (const item of payload) {
    if (!item || typeof item !== "object") {
      continue;
    }
    const type = String(item.type ?? "");
    if (!isBlockType(type)) {
      continue;
    }
    nodes.push({
      id: String(item.id ?? ""),
      type: type as BlockType,
      file: String(item.file ?? ""),
      title: String(item.title ?? ""),
      content_hash: String(item.content_hash ?? ""),
      structure_hash: String(item.structure_hash ?? "")
    });
  }
  return nodes.filter((node) => node.id.length > 0 && node.file.length > 0);
}

export async function saveNodes(sidecar: string, nodes: readonly Node[]): Promise<void> {
  const sorted = [...nodes].sort((a, b) => `${a.file}|${a.id}`.localeCompare(`${b.file}|${b.id}`));
  await writeJson(path.join(sidecar, "nodes.json"), sorted);
}

export async function loadEdges(sidecar: string): Promise<Edge[]> {
  const payload = await readJson<Array<Record<string, unknown>>>(path.join(sidecar, "edges.json"));
  if (!Array.isArray(payload)) {
    return [];
  }

  const edges: Edge[] = [];
  for (const item of payload) {
    if (!item || typeof item !== "object") {
      continue;
    }
    const type = String(item.type ?? "");
    if (!isRelationType(type)) {
      continue;
    }
    const origin = String(item.origin ?? "deterministic");
    const safeOrigin: EdgeOrigin = origin === "inferred" ? "inferred" : "deterministic";
    edges.push({
      source: String(item.source ?? ""),
      target: String(item.target ?? ""),
      type: type as RelationType,
      weight: Number(item.weight ?? 1),
      confidence: Number(item.confidence ?? 1),
      origin: safeOrigin
    });
  }
  return edges.filter((edge) => edge.source.length > 0 && edge.target.length > 0);
}

export async function saveEdges(sidecar: string, edges: readonly Edge[]): Promise<void> {
  const sorted = [...edges].sort((a, b) =>
    `${a.source}|${a.type}|${a.target}|${a.origin}`.localeCompare(`${b.source}|${b.type}|${b.target}|${b.origin}`)
  );
  await writeJson(path.join(sidecar, "edges.json"), sorted);
}
